### PR TITLE
ENH: TabularMSA.join

### LINKED ISSUE #1093: `TabularMSA`: join
From [RFC](https://github.com/biocore/scikit-bio-rfcs/blob/master/active/001-tabular-msa.md):

- `join(TabularMSA, how='strict') -> TabularMSA`

   Joins two `TabularMSA` by sequence (horizontally). `how` may be
   `'strict'`, in which case both MSAs must have the same keys (or be the same
   length if there are no keys). `how` may also be one of the common SQL joins:
   `'inner'`, `'outer'`, `'left'` (outer), or `'right'` (outer); these require
   keys or they will raise an exception. `how` is a superset of the interface
   provided by pandas. Outer based modes will insert gaps for missing keys.
   This should be documented as *not* preforming an alignment.

### PR DESCRIPTION
**Pull request by jairideout**
Fixes #1093.

### CODE REVIEW

File: skbio/alignment/_tabular_msa.py [hunk 1]
```diff
@@ -1095,6 +1095,252 @@ def _assert_valid_sequences(self, sequences):
...
+        sort
+        skbio.sequence.Sequence.concat
+
+        Notes
+        -----
+        The join operation does not automatically perform re-alignment;
+        sequences are simply joined together. Therefore, this operation is not
+        necessarily meaningful on its own.
+
+        The index of this MSA and the index of `other` must each be unique;
```
Comment (ebolyen) [inline skbio/alignment/_tabular_msa.py]: Add "respectively"? The wording as is could be construed as no matching labels between either index

File: skbio/alignment/_tabular_msa.py [hunk 2]
```diff
@@ -1095,6 +1095,252 @@ def _assert_valid_sequences(self, sequences):
...
+        if self.dtype is not other.dtype:
+            raise TypeError(
+                "`other` dtype %r does not match this MSA's dtype %r" %
+                (other.dtype if other.dtype is None else other.dtype.__name__,
+                 self.dtype if self.dtype is None else self.dtype.__name__))
+
+        if not self.index.is_unique or not other.index.is_unique:
+            raise ValueError(
+                "Cannot join `TabularMSA` objects. Index labels must be "
+                "unique.")
```
Comment (ebolyen) [inline skbio/alignment/_tabular_msa.py]: Same here, users might get the wrong impression as to how unique things have to be.

**Comment by ebolyen** [comment]
@jairideout aside from my nitpicky comment about uniqueness this looks great!

File: skbio/alignment/_tabular_msa.py (hunk 1 above)
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: I reworded and split the sentence in two, I think it reads clearer

File: skbio/alignment/_tabular_msa.py (hunk 2 above)
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: Split this conditional into two if statements to give more precise and clear error messages

**Comment by jairideout** [comment]
@ebolyen thanks for reviewing! I addressed your comments.

**Comment by ebolyen** [comment]
Thanks @jairideout!

**Pull request merged**